universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -0,0 +1,143 @@
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    Program program =
+        programDefinition.toBuilder().addBlockDefinition(blockDefinition).build().toProgram();
+    return programRepository.updateProgramSync(program).getProgramDefinition();
+  }
+
+  @Override
+  public ProgramDefinition setBlockQuestions(
+      long programId, long blockDefinitionId, ImmutableList<QuestionDefinition> questionDefinitions)
+      throws ProgramNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setQuestionDefinitions(questionDefinitions)
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  @Override
+  public ProgramDefinition setBlockHidePredicate(
+      long programId, long blockDefinitionId, Predicate predicate) throws ProgramNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setHidePredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  @Override
+  public ProgramDefinition setBlockOptionalPredicate(
+      long programId, long blockDefinitionId, Predicate predicate) throws ProgramNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setOptionalPredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  int getBlockDefinitionIndex(ProgramDefinition programDefinition, Long blockDefinitionId) {
+    return programDefinition.blockDefinitions().stream()
+        .map(b -> b.id())
+        .collect(ImmutableList.toImmutableList())
+        .indexOf(blockDefinitionId);
```

Comment:
optional suggestion: another way so that you don't have to collect all the IDs:

```java
programDefinition.blockDefinitions().stream()
    .filter(b -> b.id().equals(blockDefinitionId))
    .findFirst();
```

Reply:
I need the index. Is there a better way to get the index?